### PR TITLE
[Intel MKL] Supporting MatMul, Transpose and Softmax with BFloat16 type

### PR DESCRIPTION
**Pull request by nhasabni**
This PR enables _MklMatMul, _MklTranspose and _MklSoftmax with BFloat16 type.
Some of the changes are suggested by Clang format checker.

### CODE REVIEW

File: tensorflow/core/kernels/mkl_matmul_op.cc [hunk 1]
```diff
@@ -173,14 +173,38 @@ class MklMatMulOp : public OpKernel {
...
+    mkldnn_gemm_bf16bf16f32(ftrans[index_transb], ftrans[index_transa], &n, &m,
+                            &k, &alpha,
+                            reinterpret_cast<const mkldnn_bfloat16_t*>(b), &ldb,
+                            reinterpret_cast<const mkldnn_bfloat16_t*>(a), &lda,
+                            &beta, c_float.flat<float>().data(), &ldc);
+
+    FloatToBFloat16(c_float.flat<float>().data(), c, c_float.NumElements());
+  }
+
+// MKLDNN only supports SGEMM
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_matmul_op.cc]:
```suggestion
// MKL-DNN only supports SGEMM and bfloat16-GEMM.
```

File: tensorflow/core/kernels/mkl_matmul_op.cc [hunk 2]
```diff
@@ -173,14 +173,38 @@ class MklMatMulOp : public OpKernel {
...
 
-  // MKLDNN only supports SGEMM
+  void MklBlasGemm(OpKernelContext* ctx, bool transa, bool transb, const int m,
+                   const int n, const int k, const bfloat16* a, const int lda,
+                   const bfloat16* b, const int ldb, bfloat16* c,
+                   const int ldc) {
+    const float alpha = 1.0f;
+    const float beta = 0.0f;
+    const char* const ftrans[] = {"N", "T", "C"};
+    int index_transa = transa ? 1 : 0;
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: Let's make `index_transa` and `index_transb` `const int`.

File: tensorflow/core/kernels/mkl_matmul_op.cc [hunk 3]
```diff
@@ -173,14 +173,38 @@ class MklMatMulOp : public OpKernel {
...
+    int index_transb = transb ? 1 : 0;
+
+    Tensor c_float;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(DT_FLOAT, {m, n}, &c_float));
+
+    // MKL DNN only supports the Fortran api and requires column major while
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_matmul_op.cc]:
```suggestion
    // MKL-DNN only supports the Fortran API and requires column major while
```

File: tensorflow/core/kernels/mkl_matmul_op.cc [hunk 4]
```diff
@@ -173,14 +173,38 @@ class MklMatMulOp : public OpKernel {
...
+    // Tensorflow uses row major so we reverse the order A and B
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_matmul_op.cc]:
```suggestion
    // Tensorflow uses row major so we reverse the order of A and B.
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your PR! I have minor comments.

**Comment by nhasabni** [comment]
hi @penpornk, thanks for review. I have addressed your points. Pls take a look.

**Review by penpornk — APPROVED** [review]
Thank you for the quick fixes!

**Comment by nhasabni** [comment]
Thank you a quick review @penpornk!
